tsdb/index/tsi1/series_file.go

```diff
@@ -196,22 +187,35 @@ func (f *SeriesFile) CreateSeriesListIfNotExists(names [][]byte, tagsSlice []mod
 		}
 
 		// Move current offset to the end.
-		sz := uint64(len(buf))
-		f.n += sz
+		sz := int64(len(buf))
+		f.size += sz
 
 		// Append new key to be added to hash map after flush.
 		offsets[i] = offset
-		newKeyRanges = append(newKeyRanges, byteRange{offset, sz})
+		newKeyRanges = append(newKeyRanges, byteRange{offset, uint64(sz)})
 	}
 
 	// Flush writer.
 	if err := f.w.Flush(); err != nil {
 		return nil, err
 	}
 
-	// Add keys to hash map.
+	// Add keys to hash map(s).
 	for _, keyRange := range newKeyRanges {
-		f.hashMap.Put(f.data[keyRange.offset:keyRange.offset+keyRange.size], keyRange.offset)
+		key := f.data[keyRange.offset : keyRange.offset+keyRange.size]
+
+		f.seriesMap.inmem.Put(key, keyRange.offset)
+
+		if f.compactingSeriesMap != nil {
+			f.compactingSeriesMap.inmem.Put(key, keyRange.offset)
+		}
+	}
+
+	// Begin compaction if in-memory map is past threshold.
+	if f.seriesMap.inmem.Len() >= SeriesMapThreshold {
+		if err := f.compactSeriesMap(); err != nil {
+			return nil, err
+		}
 	}
```

Comment:
This is the blocking compaction. I'm trying to figure out the best approach here. Rebuilding the whole map requires more writes but it makes lookups faster later on. Performing incremental maps reduces write load but requires more work during lookups.